crates/bevy_ecs/src/change_detection.rs

```diff
@@ -373,6 +373,30 @@ impl<'a> MutUntyped<'a> {
     pub fn into_inner(self) -> PtrMut<'a> {
         self.value
     }
+
+    /// Turn this [`MutUntyped`] into a [`Mut`] by mapping the inner [`PtrMut`] to another value.
+    /// If you know the type of the value you can do
+    /// ```no_run
+    /// # use bevy_ecs::change_detection::{Mut, MutUntyped};
+    /// # let mut_untyped: MutUntyped = unimplemented!();
+    /// // SAFETY: ptr is of type `u8`
+    /// mut_untyped.to_typed(|ptr| unsafe { ptr.deref_mut::<u8>() });
+    /// ```
+    /// If you have a [`ReflectFromPtr`](bevy_reflect::ReflectFromPtr) that you know belongs to this [`MutUntyped`],
+    /// you can do
+    /// ```no_run
+    /// # use bevy_ecs::change_detection::{Mut, MutUntyped};
+    /// # let mut_untyped: MutUntyped = unimplemented!();
+    /// # let reflect_from_ptr: bevy_reflect::ReflectFromPtr = unimplemented!();
+    /// // SAFETY: from the context it is known that `ReflectFromPtr` was made for the type of the `MutUntyped`
+    /// mut_untyped.to_typed(|ptr| unsafe { reflect_from_ptr.as_reflect_ptr_mut(ptr) });
+    /// ```
+    pub fn to_typed<T: ?Sized>(self, f: impl FnOnce(PtrMut<'a>) -> &'a mut T) -> Mut<'a, T> {
```

Comment:
IMO this should be called `map_unchanged` (or something similar), since this is essentially the untyped equivalent of `Mut::map_unchanged`.

I would expect a function named `to_typed` to look like:

```rust
pub unsafe fn to_typed<T: ?Sized>(self) -> Mut<'a, T> {
    Mut {
        value: self.value.deref_mut(),
        ticks: self.ticks,
    }
}
```